src/pydvl/influence/torch/influence_function_model.py

```diff
@@ -351,109 +353,115 @@ def to(self, device: torch.device):
-        Compute the hessian matrix based on a provided dataloader.
-
+        Update the regularization parameter.
         Args:
-            data: The data to compute the Hessian with.
+            regularization: Either a positive float or a dictionary with the
+                block names as keys and the regularization values as values.
 
         Returns:
-            The fitted instance.
+            The modified instance
+
         """
-        self.hessian = hessian(self.model, self.loss, data)
+        self._regularization_dict = self._build_regularization_dict(regularization)
+        for k, reg in self._regularization_dict.items():
+            self.block_mapper.composable_block_dict[k].op.regularization = reg
         return self
 
-    @log_duration
-    def influences(
+    def _create_block(
         self,
-        x_test: torch.Tensor,
-        y_test: torch.Tensor,
-        x: Optional[torch.Tensor] = None,
-        y: Optional[torch.Tensor] = None,
-        mode: InfluenceMode = InfluenceMode.Up,
-    ) -> torch.Tensor:
-        r"""
-        Compute approximation of
-
-        \[ \langle H^{-1}\nabla_{\theta} \ell(y_{\text{test}},
-            f_{\theta}(x_{\text{test}})),
-            \nabla_{\theta} \ell(y, f_{\theta}(x)) \rangle, \]
-
-        for the case of up-weighting influence, resp.
-
-        \[ \langle H^{-1}\nabla_{\theta} \ell(y_{\text{test}},
-            f_{\theta}(x_{\text{test}})),
-            \nabla_{x} \nabla_{\theta} \ell(y, f_{\theta}(x)) \rangle \]
-
-        for the perturbation type influence case. The action of $H^{-1}$ is achieved
-        via a direct solver using [torch.linalg.solve][torch.linalg.solve].
-
-        Args:
-            x_test: model input to use in the gradient computations of
-                $H^{-1}\nabla_{\theta} \ell(y_{\text{test}},
-                    f_{\theta}(x_{\text{test}}))$
-            y_test: label tensor to compute gradients
-            x: optional model input to use in the gradient computations
-                $\nabla_{\theta}\ell(y, f_{\theta}(x))$,
-                resp. $\nabla_{x}\nabla_{\theta}\ell(y, f_{\theta}(x))$,
-                if None, use $x=x_{\text{test}}$
-            y: optional label tensor to compute gradients
-            mode: enum value of [InfluenceMode]
-                [pydvl.influence.base_influence_function_model.InfluenceMode]
-
-        Returns:
-            A tensor representing the element-wise scalar products for the
-                provided batch.
+        block_params: Dict[str, torch.nn.Parameter],
+        data: DataLoader,
+        regularization: Optional[float],
+    ) -> TorchOperatorGradientComposition:
+        gp = TorchGradientProvider(self.model, self.loss, restrict_to=block_params)
 
-        """
-        return super().influences(x_test, y_test, x, y, mode=mode)
+        if self.second_order_mode is SecondOrderMode.GAUSS_NEWTON:
+            mat = gauss_newton(self.model, self.loss, data, restrict_to=block_params)
```

Comment:
More a question than a comment: Is it efficient to materialize the Gauss Newton matrix and then use a direct solver? I thought it would be more efficient (and numerically robust) to build up the solutions from the factors of the Gauss Newton matrix (e.g. something like QR decomposition on the factors, triangular solve on R and then multiply).

Reply:
I created Issue #594 for this